src/topology/local_homeomorph.lean

```diff
@@ -0,0 +1,479 @@
+
+This file defines homeomorphisms between open subsets of topological spaces. An element `e` of
+`local_homeomorph α β` is an extension of `local_equiv α β`, i.e., it is a pair of functions
+`e.to_fun` and `e.inv_fun`, inverse of each other on the sets `e.source` and `e.target`.
+Additionally, we require that these sets are open, and that the functions are continuous on them.
+Equivalently, they are homeomorphisms there.
+
+Contrary to equivs, we do not register the coercion to functions and we use explicitly to_fun and
+inv_fun: coercions create unification problems for manifolds.
+
+## Main definitions
+
+`homeomorph.to_local_homeomorph`: associating a local homeomorphism to a homeomorphism, with
+                                  source = target = univ
+`local_homeomorph.symm`  : the inverse of a local homeomorphism
+`local_homeomorph.trans` : the composition of two local homeomorphisms
+`local_homeomorph.refl`  : the identity local homeomorphism
+`local_homeomorph.of_set`: the identity on a set `s`
+`eq_on_source`           : equivalence relation describing the "right" notion of equality for local
+                           homeomorphisms
+
+## Implementation notes
+
+Most statements are copied from their local_equiv versions, although some care is required
+especially when restricting to subsets, as these should be open subsets.
+
+For design notes, see `local_equiv.lean`.
+-/
+
+open function set
+
+variables {α : Type*} {β : Type*} {γ : Type*} {δ : Type*}
+[topological_space α] [topological_space β] [topological_space γ] [topological_space δ]
+
+/-- local homeomorphisms, defined on open subsets of the space -/
+structure local_homeomorph (α : Type*) (β : Type*) [topological_space α] [topological_space β]
+  extends local_equiv α β :=
+(open_source        : is_open source)
+(open_target        : is_open target)
+(continuous_to_fun  : continuous_on to_fun source)
+(continuous_inv_fun : continuous_on inv_fun target)
+
+/-- A homeomorphism induces a local homeomorphism on the whole space -/
+def homeomorph.to_local_homeomorph (e : homeomorph α β) :
+  local_homeomorph α β :=
+{ open_source        := is_open_univ,
+  open_target        := is_open_univ,
+  continuous_to_fun  := by { erw ← continuous_iff_continuous_on_univ, exact e.continuous_to_fun },
+  continuous_inv_fun := by { erw ← continuous_iff_continuous_on_univ, exact e.continuous_inv_fun },
+  ..e.to_equiv.to_local_equiv }
+
+namespace local_homeomorph
+
+variables (e : local_homeomorph α β) (e' : local_homeomorph β γ)
+
+lemma eq_of_local_equiv_eq {e e' : local_homeomorph α β}
+  (h : e.to_local_equiv = e'.to_local_equiv) : e = e' :=
+begin
+  cases e, cases e',
+  dsimp at *,
+  induction h,
+  refl
+end
+
+@[extensionality]
+protected lemma ext (e' : local_homeomorph α β) (h : ∀x, e.to_fun x = e'.to_fun x)
+  (hsymm : ∀x, e.inv_fun x = e'.inv_fun x) (hs : e.source = e'.source) : e = e' :=
+eq_of_local_equiv_eq (local_equiv.ext e.to_local_equiv e'.to_local_equiv h hsymm hs)
```

Comment:
Why do we even need to know `hsymm` here (and for `local_equiv.ext`)? Once `source` is determined, `target` is determined as the image under `to_fun`, so `inv_fun` is determined. Oh --- the problem is that `inv_fun` is only determined on `target`, and that's not the notion of equality you're using here?

If that's the explanation, perhaps adding a note reminding the reader here that the notion of equality for `local_homeomorph` is still "too strong".

Reply:
Yes, that's the reason. I have added a comment explaining this.